kafka-rest/src/main/java/io/confluent/kafkarest/ProducerMetrics.java

```diff
@@ -0,0 +1,395 @@
+        .build();
+
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.RESPONSE_SIZE_SENSOR)
+        .addAvg(
+            ProducerMetricsRegistry.RESPONSE_SIZE_AVG,
+            ProducerMetricsRegistry.RESPONSE_SIZE_AVG_DOC)
+        .build();
+
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.RECORD_ERROR_SENSOR)
+        .addRate(
+            ProducerMetricsRegistry.RECORD_ERROR_RATE,
+            ProducerMetricsRegistry.RECORD_ERROR_RATE_DOC)
+        .addWindowedCount(
+            ProducerMetricsRegistry.RECORD_ERROR_TOTAL_WINDOWED,
+            ProducerMetricsRegistry.RECORD_ERROR_TOTAL_WINDOWED_DOC)
+        .build();
+
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.REQUEST_LATENCY_SENSOR)
+        .addMax(
+            ProducerMetricsRegistry.REQUEST_LATENCY_MAX,
+            ProducerMetricsRegistry.REQUEST_LATENCY_MAX_DOC)
+        .addAvg(
+            ProducerMetricsRegistry.REQUEST_LATENCY_AVG,
+            ProducerMetricsRegistry.REQUEST_LATENCY_AVG_DOC)
+        .build();
+  }
+
+  private class BeanCoordinate {
+    private final String beanName;
+    private final Map<String, String> tags;
+
+    public BeanCoordinate(String beanName, Map<String, String> tags) {
+      this.beanName = beanName;
+      this.tags = tags;
+    }
+
+    public String getBeanName() {
+      return beanName;
+    }
+
+    public Map<String, String> getTags() {
+      return tags;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      BeanCoordinate that = (BeanCoordinate) o;
+      return Objects.equals(beanName, that.beanName) && Objects.equals(tags, that.tags);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(beanName, tags);
+    }
+  }
+
+  /**
+   * A class representing a JMX MBean where each metric maps to an MBean attribute. Sensors should
+   * be added via the {@code sensor} methods on this class, rather than directly through the {@link
+   * Metrics} class, so that the sensor names are made to be unique (based on the MBean name) and so
+   * the sensors are removed when this group is {@link #close() closed}.
+   */
+  public class ProduceMetricMBean implements AutoCloseable {
+    private final BeanCoordinate beanCoordinate;
+
+    /**
+     * Create a produce metrics MBean.
+     *
+     * @param beanCoordinate the identifier of the bean; may not be null and must be valid
+     */
+    ProduceMetricMBean(BeanCoordinate beanCoordinate) {
+      Objects.requireNonNull(beanCoordinate);
+      this.beanCoordinate = beanCoordinate;
+    }
+
+    public void recordMetrics(String sensorName, double value) {
+      Sensor sensor =
+          metrics.getSensor(
+              String.join(":", JMX_PREFIX, ProducerMetricsRegistry.GROUP_NAME, sensorName));
+      if (sensor != null) {
+        sensor.record(value);
+      }
+    }
```

Comment:
I think join is quicker than format but I get the point about concatenation every time. I've factored them out into separate methods against constants to avoid this.